packages/react/src/JsonFormsContext.tsx

```diff
@@ -126,11 +129,53 @@ const useEffectAfterFirstRender = (
   }, dependencies);
 };
 
+interface JsonFormsCoreStateProps
+  extends Omit<
+    JsonFormsInitStateProps,
+    | 'ajv'
+    | 'data'
+    | 'schema'
+    | 'uischema'
+    | 'validationMode'
+    | 'additionalErrors'
+  > {
+  core: JsonFormsCore;
+  config: Config;
+}
+interface JsonFormsStateProviderProps {
+  children: React.ReactNode;
+  initState: JsonFormsCoreStateProps;
+}
+
 export const JsonFormsStateProvider = ({
   children,
   initState,
   onChange,
-}: any) => {
+}: JsonFormsStateProviderProps & JsonFormsReactProps) => {
+  const onChangeRef = useRef(onChange);
+  useEffect(() => {
+    onChangeRef.current = onChange;
+  }, [onChange]);
+
+  const StateProvider =
+    initState.config.mode === 'controlled'
+      ? JsonFormsControlledStateProvider
+      : initState.config.mode === 'uncontrolled'
+      ? JsonFormsUncontrolledStateProvider
+      : JsonFormsMixedStateProvider;
+
+  return (
+    <StateProvider initState={initState} onChangeRef={onChangeRef}>
+      {children}
+    </StateProvider>
+  );
+};
+
+const JsonFormsMixedStateProvider = ({
```

Comment:
Maybe something you'll come back to as the PR is refined, but might there be an opportunity to repeat the amount of repeated code between `JsonFormsMixedStateProvider`, `JsonFormsControlledStateProvider` and `JsonFormsUncontrolledStateProvider` (e.g. either by having some state providers be defined in terms of other state providers, by having the providers share re-usable blocks)?